src/plugins/techStackGeneric.nim

```diff
@@ -0,0 +1,408 @@
+        if contains(excludePthRules, filePath) and contains(excludePthRules[filePath], rule_name):
+            continue
+
+        # first check if the rule should always be added for this exact file path
+        if contains(pthRules, filePath) and contains(pthRules[filePath], rule_name):
+            applicable_rules.add(rule_name)
+            continue
+
+
+        #
+        # If we don't have a filepath based, match, go by file extension and check
+        # if the extension should be getting excluded or not.
+
+        # first check if the rule applies to all file types (FT_ANY). In this
+        # case we may only have excluded filetypes, so make a pass for those,
+        # otherwise add it
+        if not contains(tsRules, rule_name):
+            continue
+
+        let tsRule = tsRules[rule_name]
+        if contains(ftRules, FT_ANY) and contains(ftRules[FT_ANY], rule_name):
+
+            # make a pass and check if we should exclude the rule
+            var exclude = false
+            if contains(ruleExcludeFiletypes, rule_name):
+                for ft in ruleExcludeFiletypes[rule_name]:
+                    # if the filetype does not match the current extension proceed
+                    if ft != splFile.ext and ft != "":
+                        continue
+                    # if we have a matching extension and a rule for that extenion,
+                    # append the rule in the rule to be run
+                    if contains(excludeFtRules, ft) and contains(excludeFtRules[ft], rule_name):
+                        exclude = true
+                        break
+            # add the rule only if its explicitly added and not excluded
+            if not exclude:
+                applicable_rules.add(rule_name)
+
+            # done processing FT_ANY
+            continue
+
+        # if the rule does not apply to all filetypes, check the ones for which
+        # it actually does apply.
+        if contains(ruleFiletypes, rule_name):
+            for ft in ruleFiletypes[rule_name]:
+                # if the filetype does not match the current extension proceed
+                if ft != splFile.ext and ft != "":
+                    continue
+
+                # if we have a matching extension and a rule for that extenion,
+                # append the rule in the rule to be run
+                if contains(ftRules, ft) and contains(ftRules[ft], rule_name):
+                    applicable_rules.add(rule_name)
+                    break
+
+    var line = ""
+    var i = 0
+    var abort = false
+    while strm.readLine(line):
+        i += 1
+        if inFileScope[category][subcategory] or abort:
+            break
+
+        for rule_name in applicable_rules:
+            if i >= headLimits[rule_name]:
+                abort = true
+                break
+
+            if find(line, regexes[rule_name]) != -1:
+                inFileScope[category][subcategory] = true
+                break
+    strm.close()
```

Comment:
if any exceptions happen above, this will never close. might want to wrap with `try..finally`, same as in python

Reply:
changing to use template for this:

```nim
template foo_(...) =
   ...
proc foo() =
   let strm = newFileStream(...)
   try:
    foo_()
  finally:
    strm.close()
```

internally compiler collapses it all into a single function but it allows for easier eading of exception handling

Reply:
complete